src/attribution/onINP.ts

```diff
@@ -180,35 +175,29 @@ const cleanupEntries = () => {
     });
   }
 
-  // The list of previous render times is used to handle cases where
-  // events are dispatched out of order. When this happens they're generally
-  // only off by a frame or two, so keeping the most recent 50 should be
-  // more than sufficient.
-  previousRenderTimes = previousRenderTimes.slice(-1 * MAX_PREVIOUS_FRAMES);
-
   // Keep all render times that are part of a pending INP candidate or
-  // that occurred within the 50 most recently-dispatched animation frames.
-  const renderTimesToKeep = new Set(
-    (previousRenderTimes as (number | undefined)[]).concat(
-      longestInteractionList.map((i) => entryToRenderTimeMap.get(i.entries[0])),
-    ),
-  );
-
-  pendingEntriesGroupMap.forEach((_, key) => {
-    if (!renderTimesToKeep.has(key)) pendingEntriesGroupMap.delete(key);
+  // that occurred within the 50 most recently-dispatched groups of events.
+  const longestInteractionGroups = longestInteractionList.map((i) => {
+    return entryToEntriesGroupMap.get(i.entries[0]);
+  });
+  const minIndex = pendingEntriesGroups.length - MAX_PREVIOUS_FRAMES;
+  pendingEntriesGroups = pendingEntriesGroups.filter((group, index) => {
+    if (index >= minIndex) return true;
+    return longestInteractionGroups.includes(group);
   });
 
   // Keep all pending LoAF entries that either:
-  // 1) intersect with entries in the newly cleaned up `pendingEntriesGroupMap`
+  // 1) intersect with entries in the newly cleaned up `pendingEntriesGroups`
   // 2) occur after the most recently-processed event entry.
   const loafsToKeep: Set<PerformanceLongAnimationFrameTiming> = new Set();
-  pendingEntriesGroupMap.forEach((group) => {
+  for (let i = 0; i < pendingEntriesGroups.length; i++) {
```

Comment:
Didn't like a `forEach()` right above a regular for loop, eh? :)

You could alternatively do something like this to keep it one line (and which is common in the Google JS style guide):

```
for (let i = 0, group; group = pendingEntriesGroups[i]; i++) {
```